src/to_substrait.cpp

```diff
@@ -853,23 +853,89 @@ substrait::Rel *DuckDBToSubstrait::TransformFilter(LogicalOperator &dop) {
 
 	if (!dfilter.projection_map.empty()) {
 		auto projection = new substrait::Rel();
-		projection->mutable_project()->set_allocated_input(res);
+		auto sproj = projection->mutable_project();
+		sproj->set_allocated_input(res);
+		auto child_column_count = GetColumnCount(*dop.children[0]);
+		auto t_index = 0;
+		vector<int32_t> output_mapping;
 		for (auto col_idx : dfilter.projection_map) {
-			CreateFieldRef(projection->mutable_project()->add_expressions(), col_idx);
+			CreateFieldRef(sproj->add_expressions(), col_idx);
+			output_mapping.push_back(child_column_count + t_index);
+			++t_index;
 		}
+		auto rel_common = CreateOutputMapping(output_mapping);
+		sproj->set_allocated_common(rel_common);
 		res = projection;
 	}
 	return res;
 }
 
+substrait::RelCommon *DuckDBToSubstrait::CreateOutputMapping(vector<int32_t> vector) {
+	auto rel_common = new substrait::RelCommon();
+	auto output_mapping = rel_common->mutable_emit()->mutable_output_mapping();
+	for (auto &col_idx : vector) {
+		output_mapping->Add(col_idx);
+	}
+	return rel_common;
+}
+
 substrait::Rel *DuckDBToSubstrait::TransformProjection(LogicalOperator &dop) {
 	auto res = new substrait::Rel();
 	auto &dproj = dop.Cast<LogicalProjection>();
+
+	auto child_column_count = GetColumnCount(*dop.children[0]);
+	auto need_output_mapping = true;
+	if (child_column_count <= dproj.expressions.size()) {
+		// check if the projection is just pass through of input columns with no reordering
+		auto exp_col_idx = 0;
+		auto is_passthrough = true;
+		for (auto &dexpr : dproj.expressions) {
```

Comment:
Consider putting this logic into a separate function to improve local readability and testing.

Reply:
done